packages/slack/src/api/base/request.ts

```diff
@@ -3,25 +3,47 @@ import * as rp from 'request-promise-native'
 import { logger, requestToLogLevel } from '../../logger'
 
 type Callback = (arg: any) => void
-type Task = { uri: string, form: any, token: string, resolve: Callback, reject: Callback }
+type Task = { uri: string, form: any, token: string, retryable: boolean, resolve: Callback, reject: Callback }
+const DEFAULT_RETRY_AFTER = 10 * 1000 // 10 seconds
+const REQUEST_TIMEOUT = Number(process.env.XENE_REQUEST_TIMEOUT) || 0
 
 const worker = async (task: Task, done) => {
   try {
     const headers = { Authorization: `Bearer ${task.token}` }
     logger.log(requestToLogLevel, `Slack API request to: ${task.uri}`)
-    await rp.post({ uri: task.uri, json: true, form: task.form, headers}).then(task.resolve)
+    await rp.post({ uri: task.uri, json: true, form: task.form, headers, timeout: REQUEST_TIMEOUT }).then(task.resolve)
+    done()
   } catch (error) {
-    if (error.statusCode !== 429) return task.reject(error)
-    const delay = Number(error.response.headers['retry-after']) * 1000
-    logger.info('Slack API rate limited for %s ms', delay)
-    setTimeout(queue.resume, delay)
-    queue.unshift(task)
-    queue.pause()
+    if (error.statusCode == 429) {
+      const delay = Number(error.response.headers['retry-after']) * 1000
+      logger.info('Slack API rate limited for %s ms', delay)
+
+      retryRequest(task, delay)
+      return done()
+    }
+
+    let retryableCodes = ['ETIMEDOUT', 'ESOCKETTIMEDOUT', 'EHOSTUNREACH', 'ECONNREFUSED']
+    if (task.retryable && error.cause && retryableCodes.includes(error.cause.code)) {
+      const delay = DEFAULT_RETRY_AFTER
+      logger.info('Slack API timeout: %s. Retrying in %s ms', error.cause.code, delay)
+
+      retryRequest(task, DEFAULT_RETRY_AFTER)
+      return done()
+    }
+    logger.error(error)
+    logger.error('Slack API request errored with status %s, timeout: %s, code: %s', error.statusCode, error.timeout, error.code)
+    return task.reject(error)
   }
-  done()
+}
+
+const retryRequest = (task: Task, delay?: number) => {
```

Comment:
```suggestion
const retryRequest = (task: Task, delayInMs?: number) => {
```

Would be helfpfull to see unit suffix in method signature